lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -276,536 +344,505 @@ bool PatchEntryPointMutate::isResourceNodeActive(const ResourceNode *node, bool
-      if ((userDataIdx % 2) != 0) {
-        *inRegMask |= 1ull << argTys.size();
-        argTys.push_back(Type::getInt32Ty(*m_context));
-        userDataIdx += 1;
+  // Add the spill table pointer and the extra args to unspilledArgs. How that is done is different between
+  // a compute shader and a graphics shader.
+  if (m_shaderStage == ShaderStageCompute) {
+    // For compute shader:
+    // If we need the spill table pointer, it must go into s11. Insert padding if necessary.
+    if (!spillTableArg.empty()) {
+      if (userDataIdx != userDataEnd) {
+        assert(userDataIdx <= userDataEnd);
+        unspilledArgs.push_back(UserDataArg(VectorType::get(builder.getInt32Ty(), userDataEnd - userDataIdx),
+                                            Util::Abi::UserDataMapping::Invalid, nullptr, /*padding=*/true));
+      }
+      unspilledArgs.push_back(spillTableArg.front());
+      // That then leaves us at a 2-aligned reg, ready for any extraUserDataArgs.
+    } else if (!extraUserDataArgs.empty()) {
+      // The extra args start with workgroupSize, which needs to start at a 2-aligned reg. Insert a single padding
+      // reg if needed.
+      if (userDataIdx & 1) {
+        unspilledArgs.push_back(
+            UserDataArg(builder.getInt32Ty(), Util::Abi::UserDataMapping::Invalid, nullptr, /*padding=*/true));
       }
-
-      auto numWorkgroupsPtrTy = PointerType::get(VectorType::get(Type::getInt32Ty(*m_context), 3), ADDR_SPACE_CONST);
-      *inRegMask |= 1ull << argTys.size();
-      entryArgIdxs.cs.numWorkgroupsPtr = argTys.size();
-      argTys.push_back(numWorkgroupsPtrTy); // NumWorkgroupsPtr
-      m_pipelineState->getPalMetadata()->setUserDataEntry(m_shaderStage, userDataIdx,
-                                                          Util::Abi::UserDataMapping::Workgroup);
-      userDataIdx += 2;
     }
-    break;
-  }
-  case ShaderStageFragment: {
-    // Do nothing
-    break;
-  }
-  default: {
-    llvm_unreachable("Should never be called!");
-    break;
-  }
-  }
-
-  if (needSpill && !useFixedLayout) {
-    *inRegMask |= 1ull << argTys.size();
-    intfData->entryArgIdxs.spillTable = argTys.size();
-    argTys.push_back(Type::getInt32Ty(*m_context));
-
-    intfData->userDataUsage.spillTable = userDataIdx;
-    m_pipelineState->getPalMetadata()->setUserDataEntry(m_shaderStage, userDataIdx++,
-                                                        Util::Abi::UserDataMapping::SpillTable);
-    intfData->spillTable.sizeInDwords = requiredUserDataCount;
+    unspilledArgs.insert(unspilledArgs.end(), extraUserDataArgs.begin(), extraUserDataArgs.end());
+  } else {
+    // For graphics shader: add the extra user data args, then the spill table pointer (if any).
+    unspilledArgs.insert(unspilledArgs.end(), extraUserDataArgs.begin(), extraUserDataArgs.end());
+    unspilledArgs.insert(unspilledArgs.end(), spillTableArg.begin(), spillTableArg.end());
   }
-  intfData->userDataCount = userDataIdx;
+}
```

Comment:
assertion about the unspilledArgs.size() at the very end here?

Reply:
What assertion are you thinking? unspilledArgs.size() is the number of unspilled args, not the number of registers (dwords).

Reply:
I hadn't realized that initially, so never mind :)